src/polydb.jl

```diff
@@ -0,0 +1,301 @@
+   return Collection{T}(db.mdb[name])
+end
+
+# search a collection for documents matching the criteria given by d
+function find(c::Collection{T}, d::Dict=Dict(); opts::Union{Nothing, Dict}=nothing) where T
+   return Cursor{T}(Mongoc.find(c.mcol, Mongoc.BSON(d); options=opts))
+end
+
+function find(c::Collection{T}, d::Pair...) where T
+   return Cursor{T}(Mongoc.find(c.mcol, Mongoc.BSON(d...)))
+end
+
+# creating `BSON` iterators from the respective `Polymake.BigObject` iterator
+function Collection{Mongoc.BSON}(c::Collection{Polymake.BigObject})
+   return Collection{Mongoc.BSON}(c.mcol)
+end
+
+function Cursor{Mongoc.BSON}(cursor::Cursor{Polymake.BigObject})
+   return Cursor{Mongoc.BSON}(cursor.mcursor)
+end
+
+# returns a Polymake.BigObject from a Mongoc.BSON document
+function parse_document(bson::Mongoc.BSON)
+   str = Mongoc.as_json(bson)
+   return call_function(:common, :deserialize_json_string, str)
+end
+
+#Iterator
+
+Base.IteratorSize(::Type{<:Cursor}) = Base.SizeUnknown()
+Base.eltype(::Cursor{T}) where T = T
+
+# help functions to reduce runtime and keep code clean
+function _evaluateIteration(nothing::Nothing)
+   return nothing
+end
+
+function _evaluateIteration(a::Tuple{Mongoc.BSON, Any})
+   return (parse_document(a[1]), a[2])
+end
+
+# default iteration functions returning `Polymake.BigObject`s
+function Base.iterate(cursor::Cursor{Polymake.BigObject})
+   return _evaluateIteration(iterate(cursor.mcursor))
+end
+
+function Base.iterate(cursor::Cursor{Polymake.BigObject}, state::Nothing)
+   return _evaluateIteration(iterate(cursor.mcursor, state))
+end
+
+function Base.iterate(coll::Collection{Polymake.BigObject})
+   return _evaluateIteration(iterate(coll.mcol))
+end
+
+function Base.iterate(coll::Collection{Polymake.BigObject}, state::Mongoc.Cursor)
+   return _evaluateIteration(iterate(coll.mcol, state))
+end
+
+# functions for `BSON` iteration
+function Base.iterate(cursor::Cursor{Mongoc.BSON})
+   return iterate(cursor.mcursor)
+end
+
+function Base.iterate(cursor::Cursor{Mongoc.BSON}, state::Nothing)
+   return iterate(cursor.mcursor, state)
+end
+
+function Base.collect(cursor::Cursor{Mongoc.BSON})
+   return collect(cursor.mcursor)
+end
+
+function Base.iterate(coll::Collection{Mongoc.BSON})
+   return iterate(coll.mcol)
+end
+
+function Base.iterate(coll::Collection{Mongoc.BSON}, state::Mongoc.Cursor)
+   return iterate(coll.mcol, state)
+end
+
+#Info
+
+# prints a list of the fields of a collection
+
+function fields(coll::Collection)
+   db = coll.mcol.database
+   coll_c = db[string("_collectionInfo.", coll.mcol.name)]
+   info = collect(coll_c)[2]
+   if haskey(info, "fields")
+      return _get_fields(info["fields"])
+   else
+      return "no information on fields available"
+   end
+end
+
+function _get_fields(d::Dict)
+   res = ""
+   for (key, value) in d
+      if value == 1
+         res = string(res, key, "\n")
```

Comment:
what does a typical field look like?
is returning always a string a good idea? a list of filtered `Pairs` would be much better i guess, something like:
`return map(p -> (last(p) == 1 ? first(p)=>nothing : p), d)`

Reply:
i now implemented a `print_fields` and a `get_fields` method. i guess, both are useful.

@benlorenz which entry from the meta collections do you recommend using? at the moment i am asking for the field "fields" of the second entry of a `_collectionInfo`, which contains either an array containing the strings of the fields, a dict with the field name string as key and a comment (or 1 if uncommented) as value or nothing at all.

Reply:
Don't use `[2]` for the info document, there is no fixed order on these documents, you should fetch them via find using the `_id`. The `fields` entry is not really widely used and maybe unstable.

I think a better (but slightly longer) way is:

1. fetch the info document `find(_collectionInfo, {_id: "info.2.1"})`
2. take the value from the `schema` field, usually just `schema.2.1`
3. fetch the corresponding schema: `find(_collectionInfo, {_id: "schema.2.1"})`
4. read the array `schema["schema"]["required"]`, these are fields that should be present in every object, maybe filter out those starting with an underscore

This should work for all collections except `SmoothReflexive`. The schema is more involved there as the list of fields depends on the dimension of the polytopes (and other properties), I'm not sure if there is an easy way to get a list of properties there. Maybe just merging all `schema["schema"]["allOf"][:]["required"]` arrays, or we could probably parse the schema with polymake and extract it from there.

PS: the 2.1 everywhere refers the current database format version.